# Web Scraping

## Top Data Analytics Companies
To improve effectiveness in business processes, companies are focussing on collecting and utilizing data. Data analytics companies enable businesses to analyze the acquired data and use them as required. Data analytics services can assist in product development, identifying potential market gaps, improving operational efficiency, etc. [Goodfirms](https://www.goodfirms.co/big-data-analytics/data-analytics)

The aim of this project is to scrape Goodfirms website for details about top data analytics company. These details includes, review, rating, year founded, location, etc.

### Road Map
Stage 1
* Download page
* Open page
* Get details
* Extract details
* Make a dataframe

Stage 2 (Code Refactoring)
* Change loops to function
* Take functions to another file
* Import functions and extract
* Make a dataframe

### Libraries

In [1]:
import requests as r
from bs4 import BeautifulSoup
import pandas as pd
import plotly.express as px
import plotly.offline as po
po.init_notebook_mode(connected=True)

### Download and save HTML file

In [2]:
# # URL link
# url = 'https://www.goodfirms.co/big-data-analytics/data-analytics'
# # access website
# html = r.get(url)
# with open('top_da_company1.html', mode='wb') as file:
#     file.write(html.content)

In [3]:
with open("top_da_company1.html", encoding='utf-8', mode='r') as file:
    bs = BeautifulSoup(file, 'lxml')
    # bs = BeautifulSoup(file, 'html5lib')

### Locate Details

In [17]:
firm_position = bs.find_all('span', {'class': "position"})
firm_names = bs.find_all('span', {'itemprop': "name"})
firm_motors = bs.find_all('p', {'class': "profile-tagline"})
firm_reviews = bs.find_all('span', {'class': "listinv_review_label"})
progress_value = bs.find_all('div', {'class': "circle-progress-value"})
firm_prices = bs.find_all('div', {'class': "firm-pricing"})
firm_emps = bs.find_all('div', {'class': "firm-employees"})
firm_years = bs.find_all('div', {'class': "firm-founded"})
firm_locations = bs.find_all('div', {'class': "firm-location"})

### Extract Details

In [19]:
pos_lst = []
for pos in firm_position:
    pos_lst.append(pos.text)

print(len(pos_lst))
pos_lst[:5]

51


['1', '2', '3', '4', '5']

In [5]:
names_lst = []
for name in firm_names[3:]:
    names_lst.append(name.text)
    
print(len(names_lst))
names_lst[:3]

51


['SPEC INDIA', 'instinctools', 'SoluLab']

In [6]:
motor_lst = []
for motor in firm_motors:
    motor_lst.append(motor.text)

print(len(motor_lst))
motor_lst[:3]

51


['Enterprise Software, Mobility & BI Solutions',
 'Delivering the future. Now.',
 'Blockchain | IoT | Mobility | AI | Big Data']

In [7]:
review_lst = []
for review in firm_reviews:
    review_lst.append(review.text)
    
print(len(review_lst))
review_lst[:3]

51


['4.8 (26 Reviews)', '4.8 (8 Reviews)', '5.0 (32 Reviews)']

In [8]:
service_pct = []
platform_pct = []
for percent in enumerate(progress_value):
    if percent[0] % 2 == 0:
        service_pct.append(percent[1].text)
    else:
        platform_pct.append(percent[1].text)
        
print(len(service_pct))
print(len(platform_pct))
print(service_pct[:3])
print(platform_pct[:3])

51
51
['20%', '5%', '15%']
['15%', '10%', '15%']


In [9]:
price_lst = []
for price in firm_prices:
    price_lst.append(price.text)
    
print(len(price_lst))
price_lst[:3]

51


['\n< $25/hr ', '\n$50 - $99/hr ', '\n$25 - $49/hr ']

In [10]:
emps_lst = []
for emp in firm_emps:
    emps_lst.append(emp.text)
    
print(len(emps_lst))
emps_lst[:3]

51


['250 - 999', '250 - 999', '50 - 249']

In [11]:
year_lst = []
for year in firm_years:
    year_lst.append(year.text)
    
print(len(year_lst))
year_lst[:3]

51


['1987', '2000', '2014']

In [12]:
firm_lst = [firm.text for firm in firm_locations]
print(len(firm_lst))
firm_lst[:3]

51


['\nIndia, United States ',
 '\nUnited States, Germany ',
 '\nUnited States, India ']

### Put Details in a DataFrame

In [20]:
df = pd.DataFrame()

In [21]:
df['firm_position'] = pos_lst
df['firm_name'] = names_lst
df['firm_motor'] = motor_lst
df['firm_review'] = review_lst
df['service_pct'] = service_pct
df['platform_pct'] = platform_pct
df['firm_price'] = price_lst
df['firm_employee'] = emps_lst
df['year_founded'] = year_lst
df['firm_location'] = firm_locations

In [22]:
# df
# df2 = df.copy()
# pd.concat([df, df2])

Task
Extract the remaining three details
Download the second page and save it as "top_da_company2.html"
Extract the same details as seen above
Create a dataframe 

In [38]:
# URL link
url = 'https://www.goodfirms.co/big-data-analytics/data-analytics?page=2'
# access website
html = r.get(url)
with open('top_da_company2.html', mode='wb') as file:
    file.write(html.content)

In [39]:
with open("top_da_company2.html", encoding='utf-8', mode='r') as file:
    bs = BeautifulSoup(file, 'lxml')
    # bs = BeautifulSoup(file, 'html5lib')

In [40]:
firm_names = bs.find_all('span', {'itemprop': "name"})
firm_motors = bs.find_all('p', {'class': "profile-tagline"})
firm_reviews = bs.find_all('span', {'class': "listinv_review_label"})
progress_value = bs.find_all('div', {'class': "circle-progress-value"})
firm_prices = bs.find_all('div', {'class': "firm-pricing"})
firm_emps = bs.find_all('div', {'class': "firm-employees"})
firm_years = bs.find_all('div', {'class': "firm-founded"})
firm_locations = bs.find_all('div', {'class': "firm-location"})

In [41]:
import functions as fn

In [42]:
fn.extract_names(firm_names)[:5]

The number of name extracted is 51


0                           Systango Technologies
1                                         Inoxoft
2    Noventum Custom Software Development Company
3                   Napollo Software Design L.L.C
4                        SemiDot Infotech Pvt Ltd
dtype: object

In [49]:
fn.extract_motors(firm_motors)[:5]

The number of firm motor extracted is 51


0                    We Make the Impossible, Possible!
1    IEC/ISO 27001, Google, Microsoft Certified Com...
2      Custom Software & Web Development in New Mexico
3      Best Software Design Agency in Dubai & New York
4            Right Technology Partner for IT Solutions
dtype: object

In [26]:
fn.extract_reviews(firm_reviews)[:5]

The number of review extracted is 51


0    4.8 (26 Reviews)
1     4.8 (8 Reviews)
2    5.0 (32 Reviews)
3     4.7 (5 Reviews)
4     5.0 (5 Reviews)
dtype: object

In [27]:
fn.extract_progress_values(progress_value)[0][:5]

The number of service percent is 51
The number of platform percent is 51


0    20%
1     5%
2    15%
3    40%
4    25%
dtype: object

In [28]:
fn.exttract_prices(firm_prices)[:5]

The number of price extracted is 51


0        \n< $25/hr 
1    \n$50 - $99/hr 
2    \n$25 - $49/hr 
3    \n$25 - $49/hr 
4    \n$25 - $49/hr 
dtype: object

In [29]:
fn.extract_employees(firm_emps)[:5]

The number of employees extracted is 51


0    250 - 999
1    250 - 999
2     50 - 249
3    250 - 999
4      10 - 49
dtype: object

In [30]:
fn.extract_founded_year(firm_years)[:5]

The number of year extracted is 51


0    1987
1    2000
2    2014
3    2010
4    2016
dtype: object

In [31]:
fn.extract_locations(firm_locations)[:5]

The number of locations extracted is 51


0        \nIndia, United States 
1      \nUnited States, Germany 
2        \nUnited States, India 
3    \nUnited States, Australia 
4        \nUnited States, India 
dtype: object

### Data Cleaning

In [23]:
df.head(5)

firm_position           firm_name  \
0             1          SPEC INDIA   
1             2        instinctools   
2             3             SoluLab   
3             4  Sigma Data Systems   
4             5       NeenOpal Inc.   

                                     firm_motor       firm_review service_pct  \
0  Enterprise Software, Mobility & BI Solutions  4.8 (26 Reviews)         20%   
1                   Delivering the future. Now.   4.8 (8 Reviews)          5%   
2   Blockchain | IoT | Mobility | AI | Big Data  5.0 (32 Reviews)         15%   
3       Discover the world of Big Data with us!   4.7 (5 Reviews)         40%   
4            The Hub Of Data Science Innovation   5.0 (5 Reviews)         25%   

  platform_pct       firm_price firm_employee year_founded  \
0          15%      \n< $25/hr      250 - 999         1987   
1          10%  \n$50 - $99/hr      250 - 999         2000   
2          15%  \n$25 - $49/hr       50 - 249         2014   
3          10%  \n$25 - $49/hr      250 - 999         2010   
4          10%  \n$25 - $49/hr        10 - 49         2016   

                   firm_location  
0      [\nIndia, United States ]  
1    [\nUnited States, Germany ]  
2      [\nUnited States, India ]  
3  [\nUnited States, Australia ]  
4      [\nUnited States, India ]

In [24]:
df.insert(2, 'star_rating', df['firm_review'].apply(lambda x: x.split()[0]))

In [25]:
df.insert(3, 'review', df['firm_review'].apply(lambda x: x.split()[1].strip('(')))

In [26]:
df.rename(columns={'service_pct':'service_pct(%)'}, inplace=True)
df['service_pct(%)'] = df['service_pct(%)'].apply(lambda x: x.strip('%'))

In [27]:
df.rename(columns={'platform_pct':'platform_pct(%)'}, inplace=True)
df['platform_pct(%)'] = df['platform_pct(%)'].apply(lambda x: x.strip('%'))

In [28]:
df['firm_price'] = df['firm_price'].apply(lambda x: x.strip('\n'))

In [29]:
df['firm_location'] = df['firm_location'].apply(lambda x: x.text.strip('\n'))

In [30]:
df.drop(columns='firm_review', inplace=True)

In [31]:
df.head(5)

firm_position           firm_name star_rating review  \
0             1          SPEC INDIA         4.8     26   
1             2        instinctools         4.8      8   
2             3             SoluLab         5.0     32   
3             4  Sigma Data Systems         4.7      5   
4             5       NeenOpal Inc.         5.0      5   

                                     firm_motor service_pct(%)  \
0  Enterprise Software, Mobility & BI Solutions             20   
1                   Delivering the future. Now.              5   
2   Blockchain | IoT | Mobility | AI | Big Data             15   
3       Discover the world of Big Data with us!             40   
4            The Hub Of Data Science Innovation             25   

  platform_pct(%)     firm_price firm_employee year_founded  \
0              15      < $25/hr      250 - 999         1987   
1              10  $50 - $99/hr      250 - 999         2000   
2              15  $25 - $49/hr       50 - 249         2014   
3              10  $25 - $49/hr      250 - 999         2010   
4              10  $25 - $49/hr        10 - 49         2016   

               firm_location  
0      India, United States   
1    United States, Germany   
2      United States, India   
3  United States, Australia   
4      United States, India

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   firm_position    51 non-null     int32  
 1   firm_name        51 non-null     object 
 2   star_rating      51 non-null     float64
 3   review           51 non-null     int32  
 4   firm_motor       51 non-null     object 
 5   service_pct(%)   51 non-null     int32  
 6   platform_pct(%)  51 non-null     int32  
 7   firm_price       51 non-null     object 
 8   firm_employee    51 non-null     object 
 9   year_founded     51 non-null     object 
 10  firm_location    51 non-null     object 
dtypes: float64(1), int32(4), object(6)
memory usage: 3.7+ KB


In [92]:
df['firm_position'] = df['firm_position'].astype('int')
df['star_rating'] = df['star_rating'].astype('float')
df['review'] = df['review'].astype('int')
df['service_pct(%)'] = df['service_pct(%)'].astype('int')
df['platform_pct(%)'] = df['platform_pct(%)'].astype('int')

### EDA

In [98]:
top5_sta_rev = df.sort_values(by=['star_rating', 'review'], ascending=False)[:5]
top5_sta_rev

firm_position firm_name  star_rating  review  \
2               3   SoluLab          5.0      32   
31             32     S-PRO          5.0      30   
35             36   MobiDev          5.0      16   
39             40  APPSTIRR          5.0      14   
41             42     7EDGE          5.0      14   

                                           firm_motor  service_pct(%)  \
2         Blockchain | IoT | Mobility | AI | Big Data              15   
31                                 Strategic partner.              10   
35                Web and Mobile development, AI apps              10   
39              Award Winning App Development Company               5   
41  Software and Product Development | Dedicated T...               5   

    platform_pct(%)     firm_price firm_employee year_founded  \
2                15  $25 - $49/hr       50 - 249         2014   
31               25  $50 - $99/hr       50 - 249         2014   
35               10  $50 - $99/hr      250 - 999         2009   
39               20  $25 - $49/hr       50 - 249         2010   
41                5  $25 - $49/hr       50 - 249         2010   

                           firm_location  
2                  United States, India   
31                               Poland   
35                        United States   
39  United States, United Arab Emirates   
41                 India, United States

In [112]:
px.histogram(df, 'star_rating', width=500, title='Star Rating Distribution')

In [113]:
px.histogram(df, 'review', width=500, title='Review Distribution')

In [101]:
px.bar(top5_sta_rev, 'firm_name', ['star_rating', 'review'], width=700)

In [104]:
top5_sta_ser = df.sort_values(by=['star_rating', 'service_pct(%)'], ascending=False)[:5]
top5_sta_ser

firm_position    firm_name  star_rating  review  \
8               9     Datapine          5.0       3   
24             25     SetuServ          5.0       1   
9              10   Dataforest          5.0       4   
20             21       VITech          5.0       3   
25             26  InData Labs          5.0       1   

                                       firm_motor  service_pct(%)  \
8                 BUSINESS INTELLIGENCE MADE EASY              90   
24  Customized AI: Machine Power + Human Accuracy              50   
9    Data Engineering and Web Product Development              30   
20                Software development powerhouse              30   
25                     Big on Data Science and AI              30   

    platform_pct(%)      firm_price firm_employee year_founded  \
8                 5   $50 - $99/hr        10 - 49         2012   
24               16             NA       50 - 249         2012   
9                20   $50 - $99/hr       50 - 249         2018   
20               25  $50 - $99/hr        50 - 249         2007   
25               30   $50 - $99/hr       50 - 249         2014   

            firm_location  
8                Germany   
24  United States, India   
9       Ukraine, Estonia   
20               Ukraine   
25     Cyprus, Singapore

In [140]:
px.bar(top5_sta_ser, 'firm_name', ['star_rating', 'service_pct(%)'], width=600)

In [123]:
df.sort_values('year_founded')[:5]

firm_position            firm_name  star_rating  review  \
0               1           SPEC INDIA          4.8      26   
1               2         instinctools          4.8       8   
33             34  CDN Solutions Group          4.6      20   
46             47           HQSoftware          4.9      11   
17             18            OrbitSoft          4.9       3   

                                           firm_motor  service_pct(%)  \
0        Enterprise Software, Mobility & BI Solutions              20   
1                         Delivering the future. Now.               5   
33  Blockchain, NFT, Mobile App, Software Development              10   
46       IoT, AR, VR, and custom software development               5   
17                               We do software right               5   

    platform_pct(%)     firm_price firm_employee year_founded  \
0                15      < $25/hr      250 - 999         1987   
1                10  $50 - $99/hr      250 - 999         2000   
33               20      < $25/hr      250 - 999         2000   
46               25  $25 - $49/hr       50 - 249         2001   
17               20  $25 - $49/hr       50 - 249         2001   

              firm_location  
0     India, United States   
1   United States, Germany   
33    India, United States   
46  United States, Georgia   
17           United States

In [139]:
df['firm_price'].value_counts()

$25 - $49/hr       27
$50 - $99/hr       11
< $25/hr            6
$50 - $99/hr        3
NA                  3
$100 - $149/hr      1
Name: firm_price, dtype: int64

In [132]:
split = df['firm_location'].apply(lambda x: x.split(', '))
country_lst = []
for lst in split:
    country_lst += lst
new = []
for country in country_lst:
    new.append(country.strip())    

countries_cnt = pd.Series(new).value_counts()

In [134]:
px.bar(x=countries_cnt.values, y=countries_cnt.index)

### Summary

* The webpage from Goodfirm was succesfully accessed and downloaded using Request library. 
* BeautifulSoup was used to locate and extract the details from the downloaded html file
* The extracted details was converted to data frame using Pandas.
* The file was cleaned and changed to its right datatype
* The cleaned file was explored for some insights.

#### Findings
* About 30 firms are rated 5 star while 12 are rated between 4.8 - 4.9
* Very few firms (3) has a review above 30
* Most of the firms are located in United States followed by India.